### PR TITLE
Fix burr() distribution stats() method and add other methods

### PR DESCRIPTION
**Pull request by rlucas7**
closes scipy issue #9544 (mean and stats method of burr() return erroneous results). 

Also fixes the issues in `fisk` and `burr12` that inherit from `burr`. 

Adds unit test(s) for `c<1` case (mean/variance are `np.nan` and `c>1` case (mean is finite). 

Also cleans up the docstring for `burr` and adds refs.

### CODE REVIEW

**Comment by rlucas7** [comment]
The pypy3 circle CI build fails are something in the signal library (no signal code touched in this PR) so unrelated.

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -762,15 +762,29 @@ class burr_gen(rv_continuous):
 
     `burr` takes :math:`c` and :math:`d` as shape parameters.
 
+    The kth moment of `burr` is:
```
[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: removed from docstring.

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -781,13 +795,46 @@ def _pdf(self, x, c, d):
...
+    def _stats(self, c, d):
+        e = lambda k: sc.beta(d + k / c, 1.0 - k / c) / sc.beta(d, 1.0)
+        e1 = e(1.0)
+        e2 = e(2.0)
+        e3 = e(3.0)
+        e4 = e(4.0)
+        mu = np.where(c > 1.0, e1, np.nan)
+        mu2_if_c = e2 - mu**2
+        mu2 = np.where(c > 2.0, mu2_if_c, np.nan)
+        if c > 3.0:
```
[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: When I put the logic for `g1` inside an `np.where()` function I get a runtime warning: 
`__main__:3: RuntimeWarning: invalid value encountered in double_scalars`
which causes the `c=0.5` and `d=3` case to fail the unit test. It seems like np.where() should only execute the logical expression required but it appears to evaluate both expressions and then decide which to return. The runtime error is from the `g1` expression trying to do arithmetic on `np.nan`.

[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: so the usecase for `np.where()` seems to be subsetting numpy vectors via indices, this explains why the eqns are evaluated whether they should e returned or not. I thought about using `_lazywhere()` in `_lib/_utils.py` but that util function has similar behavior (evaluates both regardless of condition and then returns only one or the other). Leaving as is for now.

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -762,15 +762,29 @@ class burr_gen(rv_continuous):
...
+
+        E(X^k) = \frac{B(d + k/c, 1 - k/c)}{B(d,1)}
+
+    where :math:`c` and :math:`d` are shape parameters and `c > k`.
+
     This is the PDF corresponding to the third CDF given in Burr's list;
-    specifically, it is equation (11) in Burr's paper [1]_.
+    specifically, it is equation (11) in Burr's paper [1]_. The distribution
+    is also commonly referred to as the Dagum distribution [2]_. If the
+    parameter :math:`c < 1` then the mean of the distribution does not
```
[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: >I only meant the part related to the mean. You can also leave it there, i personally do not find it > necessary

Given that the distribution is less well known and the mean/variance are `np.nan` when the distribution has `c<1` (`c<2` respectively). I think it makes sense to keep in docstring to inform the casual user. If someone has a reason to omit from docstring, I would take it out though.

**Comment by rlucas7** [comment]
looks like the 2 failures are random; circile ci "Too long with no output (exceeded 10m0s)" and the azure build is related to mpmath precision issues at complex values of z. Both having nothing to do w/my changes. I'm going to remove the LIC44-45 of `tests/test_continuous_basic.py` and see if the build still works. Other than these changes I'll just address whatever comments there are in the PR that come up (I think I got all of @chrisb83 comments but let me know if I missed something).  

># burr: sample mean test fails still for c<1
>?['burr', (0.94839838075366045, 4.3820284068855795)],

File: scipy/stats/_continuous_distns.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: @ev-br you are correct, looks like I read
https://github.com/scipy/scipy/blob/master/scipy/_lib/_util.py#L29-L31
too hastily. I'll swap in the _lazywhere here for `g1` and `g2` calculations. 

Thanks for catching!

File: scipy/stats/_continuous_distns.py [hunk 4]
```diff
@@ -781,13 +787,47 @@ def _pdf(self, x, c, d):
...
+
     def _ppf(self, q, c, d):
         return (q**(-1.0/d) - 1)**(-1.0/c)
 
+    def _stats(self, c, d):
+        e = lambda k: sc.beta(d + k / c, 1.0 - k / c) / sc.beta(d, 1.0)
+        e1 = e(1.0)
+        e2 = e(2.0)
+        e3 = e(3.0)
+        e4 = e(4.0)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Nit: sc.beta(d + np.arange(1, 5)/c, 1. - np.arange(1, 5)/c)`

**Comment by rlucas7** [comment]
> Denominator is missing?

Too late at night I guess :). Most recent pushed commit should resolve the issues (except for flaky tests). 

Thanks for catching! I couldn't figure out what the issue was last night.

**Comment by rlucas7** [comment]
For the Azure build the log shows the same tests as before; 

\special\tests\test_mpmath.py::TestSystematic

tests under this are nothing I've touched, not sure if this is flaky b/c of windows 32 bit machine or just windows :). 

Everything else looks good.

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: in most recent commit.

File: scipy/stats/_continuous_distns.py [hunk 5]
```diff
@@ -781,13 +795,46 @@ def _pdf(self, x, c, d):
...
+            g2 = (e4 - 4 * e3 * mu + 6 * e2 * mu**2 - 3 * mu**4) /\
+                    mu2_if_c**2
+        else:
+            g2 = np.nan
+        g1 = np.where(c > 3.0, g1, np.nan)
+        g2 = np.where(c > 4.0, g2, np.nan)
+        return mu, mu2, g1, g2
+
     def _munp(self, n, c, d):
+        # TODO: figure out if we still need this in light of _stats()
```
[the remark this answers is not in the document]
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: removedf the #TODO in most recent commit.

**Comment by rlucas7** [comment]
@ev-br @chrisb83 @WarrenWeckesser @josef-pkt friendly bump, anything else I should change on this PR?

File: scipy/stats/_continuous_distns.py [hunk 6]
```diff
@@ -781,12 +787,39 @@ def _pdf(self, x, c, d):
         # burr.pdf(x, c, d) = c * d * x**(-c-1) * (1+x**(-c))**(-d-1)
         return c * d * (x**(-c - 1.0)) * ((1 + x**(-c))**(-d - 1.0))
 
+    def _logpdf(self, x, c, d):
+        return np.log(c) + np.log(d) + (-c - 1.0) * np.log(x) \
```
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py]: Don't use a backslash for line continuation.  From [the section of PEP 8 on line length](https://www.python.org/dev/peps/pep-0008/#maximum-line-length):

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

File: scipy/stats/_continuous_distns.py [hunk 7]
```diff
@@ -781,12 +787,39 @@ def _pdf(self, x, c, d):
...
+                + (-d - 1.0) * np.log1p(x**(-c))
+
     def _cdf(self, x, c, d):
         return (1 + x**(-c))**(-d)
 
+    def _logcdf(self, x, c, d):
+        return -d * np.log1p(x**(-c))
+
+    def _sf(self, x, c, d):
```
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py]: This implementation of `_sf` is just `1 - cdf(x)`, so there is no need to include it.  That's what the default method does.  (There is a different formulation that is better at handling the right tail that might be worth implementing, but I'll experiment a bit and open a separate pull request if it looks worthwhile.)

File: scipy/stats/_continuous_distns.py [hunk 8]
```diff
@@ -781,12 +787,39 @@ def _pdf(self, x, c, d):
...
+        return 1 - (1 + x**(-c))**(-d)
+
     def _ppf(self, q, c, d):
         return (q**(-1.0/d) - 1)**(-1.0/c)
 
+    def _stats(self, c, d):
+        nc = np.arange(1, 5)/c
+        e1, e2, e3, e4 = sc.beta(d + nc, 1. - nc) / sc.beta(d, 1.0)
```
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py]: Simplify: `beta(d, 1.0)` is just `1/d`, so this expression can be written `d * sc.beta(d + nc, 1. - nc)`.

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
Thanks @rlucas7.  I made some comments in-line.  Here's one more.

The `mean` and `stats` methods should handle broadcasting, but that does work in this pull request:

```
IIn [29]: burr.mean([2, 3], 5)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-29-b4c0faf546c1> in <module>()
----> 1 burr.mean([2, 3], 5)

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+283ebdf-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_distn_infrastructure.py in mean(self, *args, **kwds)
   1215         """
   1216         kwds['moments'] = 'm'
-> 1217         res = self.stats(*args, **kwds)
   1218         if isinstance(res, ndarray) and res.ndim == 0:
   1219             return res[()]

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+283ebdf-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_distn_infrastructure.py in stats(self, *args, **kwds)
   1011                                               **{'moments': moments})
   1012             else:
-> 1013                 mu, mu2, g1, g2 = self._stats(*goodargs)
   1014             if g1 is None:
   1015                 mu3 = None

~/miniconda36/lib/python3.6/site-packages/scipy-1.3.0.dev0+283ebdf-py3.6-macosx-10.7-x86_64.egg/scipy/stats/_continuous_distns.py in _stats(self, c, d)
    805 
    806     def _stats(self, c, d):
--> 807         nc = np.arange(1, 5)/c
    808         e1, e2, e3, e4 = sc.beta(d + nc, 1. - nc) / sc.beta(d, 1.0)
    809         mu = np.where(c > 1.0, e1, np.nan)

ValueError: operands could not be broadcast together with shapes (4,) (2,) 
```

Here's the behavior from the master branch, for `burr` and a couple other distributions:

```
In [18]: burr.mean([2, 3], 5)
Out[18]: array([3.86563159, 2.26413912])

In [19]: norm.stats([0, 1, 2], [[0.1], [10]])
Out[19]: 
(array([[0., 1., 2.],
        [0., 1., 2.]]), array([[1.e-02, 1.e-02, 1.e-02],
        [1.e+02, 1.e+02, 1.e+02]]))

In [20]: gamma.stats([[1], [2]], loc=[3, 4])
Out[20]: 
(array([[4., 5.],
        [5., 6.]]), array([[1., 1.],
        [2., 2.]]))
```

`burr.stats` in the current master branch doesn't handle broadcasting either... 

```
In [21]: burr.stats([2, 3], 5)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-21-ec439142bfe8> in <module>()
----> 1 burr.stats([2, 3], 5)

~/miniconda36b/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py in stats(self, *args, **kwds)
   1036                         mu = self._munp(1, *goodargs)
   1037                     mu2 = mu2p - mu * mu
-> 1038                     if np.isinf(mu):
   1039                         # if mean is inf then var is also inf
   1040                         mu2 = np.inf

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
so it will be nice to get that fixed in this pull request.

File: scipy/stats/_continuous_distns.py (hunk 6 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: done, added parens around entire expression.

File: scipy/stats/_continuous_distns.py (hunk 7 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: ok, removed `_sf`

File: scipy/stats/_continuous_distns.py (hunk 8 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: right, same as `_munp`, done.

**Comment by rlucas7** [comment]
@WarrenWeckesser Thanks for the comments! I hadn't caught the broadcast error, good eyes! 
I reshaped the `np.arange(1,5)` to `np.arange(1,5).reshape(4,1)` which appears to fix the broadcasting issue for `_stats` and `mean`. I think this most recent commit (commit <5 minutes ago) addresses the 3 comments you've made here. Please let me know if I missed something.

File: scipy/stats/_continuous_distns.py (hunk 7 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: So removing `_sf` caused some errors in the unit tests for `fisk`. My solution was to reimplement `_sf` in `fisk` w/ `d=1.0`. The error message was too many arguments getting passed from burr to fisk, this came up on stats method unit test too.

**Comment by rlucas7** [comment]
>I think this most recent commit (commit <5 minutes ago) addresses the 3 comments you've made 

Nope, the broadcast wasn't fixed. Most recent commit fixes broadcast and hacks to fix issues that propogate from (I think) removal of `_sf`.

**Comment by rlucas7** [comment]
> > I think this most recent commit (commit <5 minutes ago) addresses the 3 comments you've made
> 
> Nope, the broadcast wasn't fixed. Most recent commit fixes broadcast and hacks to fix issues that 
> propogate from (I think) removal of `_sf`.

Nope. I put `_sf` implementation back into `burr` and ran unit tests that doesn't fix the error I was getting. However, the most recent commit fixes both broadcast issue and the too many args issue in `fisk`. @WarrenWeckesser your comments have been addressed and passing CI, (azure build is flaky), do you have more comments for me to address?

**Comment by WarrenWeckesser** [comment]
@rlucas7, thanks.  I probably won't get back to this until after the holidays.

**Comment by rlucas7** [comment]
@WarrenWeckesser the most recent push removes some of the Fisk distribution changes (they weren't necessary after some cleanup) and addresses the broadcast issue. The method to fix the broadcast unwraps the first 4 moments into separate lines of code. This works and I can't think of a more elegant way to do the first 4 moments. Your code is more elegant than mine so let me know if you have any idea on a fix. Otherwise, I'll leave as is.

**Comment by rlucas7** [comment]
git rebase fail, git 1, rlucas7, 0. :(

**Comment by ev-br** [comment]
Relevant commits are 5770438 and below, correct? I'd cherry-pick them into a new branch, then force-push here.

**Comment by rlucas7** [comment]
@ev-br thanks for suggesting the approach, not clear to me if by "here" you mean the pull request branch directly or the associated branch on my forked repo. Could you clarify which is meant by 
"here"?

**Comment by ev-br** [comment]
Here I squashed your 30+ commits to make cherry-picking easier, please check.
https://github.com/ev-br/scipy/pull/new/pr/9585-copy

A way to squash is to first `git log` to find the first commit beyond your commits, then
`git rebase -i 74c32fb2914bb76e90a91929b03ede6330dfbfcf` where the commit hash is what you've just found. Then in the editor which opens, change `pick` to `f`, close the editor. If there are conflicts, resolve them (`git status`, open the files which are "both" changed, search for `>>>>>>>`, edit, save, `git add <filename> && git rebase --continue`)

Notice that I did all this in a copy of a branch --- best make a copy whenever you're about to rebase things.

Once there's a single commit then just `git cherry-pick <commit hash>` , test it, and `git push origin --force <pr-branch-name>`.

Two main gotchas are 
- interactive rebase off too old commit (I guess this is what happened here). Always find the most recent commit for which to rebase onto.
- force-pushing to the remote repo (`git push --force`). This is needed because rebasing changes commit hashes.

Hope this helps :-)

**Comment by rlucas7** [comment]
@ev-br Thanks for the helpful description of rebase. I had 2 gotchas, 1 which you identified, I had an off by 1 on the SHA codes in the `git log` and second I was using `git rebase` and *NOT* `git rebase -i` the interactive option helps clarify what is going on under the git hood, thanks! 

However, I've still stuck. I got all the way to a few minor differences between the origin branch and the branch on my local and want to do a force push and it still won't let me. 

```
(scipy-dev) Lucass-MacBook:stats rlucas$ git remote -vv
origin	https://github.com/rlucas7/scipy.git (fetch)
origin	https://github.com/rlucas7/scipy.git (push)
upstream	git://github.com/scipy/scipy.git (fetch)
upstream	git://github.com/scipy/scipy.git (push)
(scipy-dev) Lucass-MacBook:stats rlucas$ git branch -vv
* gh_issue_9544    190cbd0ac WIP: rebase fail fixup
  irwinhall        bdcb4fbd9 TST-WIP: code to test Irwin-Hall; needs unit tests?
  master           beea19898 [origin/master] Merge pull request #9931 from hroskes/master
  new_branch_temp  fb9a99659 WIP: code to :quote:`fix` stats
  pr/9539          f04fbcdc4 [upstream/pr/9539: ahead 3, behind 374] return named tuple; minor doc update
  scipy_issue_9544 8321988e3 [origin/scipy_issue_9544] Rebase 'scipy_issue_9544' of https://github.com/rlucas7/scipy with upstream master
(scipy-dev) Lucass-MacBook:stats rlucas$ git diff gh_issue_9544 origin/scipy_issue_9544
diff --git a/scipy/stats/_continuous_distns.py b/scipy/stats/_continuous_distns.py
index 9727f8554..3da4601a5 100644
--- a/scipy/stats/_continuous_distns.py
+++ b/scipy/stats/_continuous_distns.py
@@ -805,8 +805,10 @@ class burr_gen(rv_continuous):
         return (q**(-1.0/d) - 1)**(-1.0/c)
 
     def _stats(self, c, d):
-        nc = np.arange(1, 5).reshape(4,1) / c
-        e1, e2, e3, e4 = d * sc.beta(d + nc, 2. - nc)
+        e1 = self._munp(1, c, d)
+        e2 = self._munp(2, c, d)
+        e3 = self._munp(3, c, d)
+        e4 = self._munp(4, c, d)
         mu = np.where(c > 1.0, e1, np.nan)
         mu2_if_c = e2 - mu**2
         mu2 = np.where(c > 2.0, mu2_if_c, np.nan)
@@ -817,7 +819,7 @@ class burr_gen(rv_continuous):
         g2 = _lazywhere(c > 3.0, (c, e1, e2, e3, e4, mu2_if_c),
              lambda c, e1, e2, e3, e4, mu2_if_c:
              ((e4 - 4 * e3 * e1 + 6 * e2 * e1**2 - 3 * e1**4) /
-             mu2_if_c**2) - 3,
+             mu2_if_c**2) - 3.,
              fillvalue=np.nan)
         return mu, mu2, g1, g2
 
@@ -976,9 +978,6 @@ class fisk_gen(burr_gen):
              fillvalue=np.nan)
         return mu, mu2, g1, g2
 
-    def _stats(self, c):
-        return burr_gen._stats(self, c, 1.0)
-
     def _entropy(self, c):
         return 2 - np.log(c)
 
(scipy-dev) Lucass-MacBook:stats rlucas$ git push --force origin scipy_issue_9544
Everything up-to-date
```

Note that I deleted the (local only) branch from before that contains rebase fail WIP and started with a new branch under the same name `gh_issue_9544`. The `Everything up-to-date` message is definitely not correct. I'm running 
```
(scipy-dev) Lucass-MacBook:stats rlucas$ git --version
git version 2.17.2 (Apple Git-113)
```
so the version is less than 1 year old. I'm still not clear why I cannot force the push onto my own remote `origin`, any ideas? (Apologies for the long message)

**Comment by ev-br** [comment]
I'd try 1) check `git status` locally, 2) check the `origin` fork (have you actually pushed it already?), 3) `git push origin local_branch:new_name`

**Comment by rlucas7** [comment]
> I'd try 1) check `git status` locally, 2) check the `origin` fork (have you actually pushed it already?), 3) `git push origin local_branch:new_name`

This creates a new branch on my forked repository with the option to open a new pull request-was that the desired outcome?

```
scipy-dev) Lucass-MacBook:stats rlucas$ git push origin scipy_issue_9544:gh_issue_9544
Total 0 (delta 0), reused 0 (delta 0)
remote: 
remote: Create a pull request for 'gh_issue_9544' on GitHub by visiting:
remote:      https://github.com/rlucas7/scipy/pull/new/gh_issue_9544
remote: 
To https://github.com/rlucas7/scipy.git
 * [new branch]          scipy_issue_9544 -> gh_issue_9544
```

**Comment by ev-br** [comment]
Yes :-). That, or delete the old local branch, rename the new one to an old name and force-push.

**Comment by rlucas7** [comment]
> Yes :-). That, or delete the old local branch, rename the new one to an old name and force-push.

Ok. I've got a new PR branch, I'll keep this one open for now until the PR is merged.

**Comment by rgommers** [comment]
@rlucas7 can we close this PR to get it out of the review queue?

**Comment by rlucas7** [comment]
> @rlucas7 can we close this PR to get it out of the review queue?

Yes closing. I had left it open thinking I would merge the new PR, but the PR from @pvanmulbregt incorporates some of this PR and other stuff too and should be the merged one. I'll close both of my PRs on this issue.

**Comment by rlucas7** [comment]
To be clear, closing in favor of PR #9983

**Pull request closed without merging**